From 7234032294158956c7bd2fa6f48e72b29cba644e Mon Sep 17 00:00:00 2001
From: "Lu, Xingjiang" <xingjiang.lu@intel.com>
Date: Mon, 4 Mar 2024 18:31:12 +0800
Subject: [PATCH] Use a private thread to replace the public one in installd
 connection. To prevent the public background thread blocks the installd
 reconnection if it is busy in initialization.

Tracked-on: OAM-116189
Signed-off-by: Xie, Hongcheng <hongcheng.xie@intel.com>
Signed-off-by: Lu, Xingjiang <xingjiang.lu@intel.com>
---
 .../core/java/com/android/server/pm/Installer.java  | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/pm/Installer.java b/services/core/java/com/android/server/pm/Installer.java
index 13c5bb0230e2..fc9d6b784e51 100644
--- a/services/core/java/com/android/server/pm/Installer.java
+++ b/services/core/java/com/android/server/pm/Installer.java
@@ -178,9 +178,16 @@ public class Installer extends SystemService {
             }
         } else {
             Slog.w(TAG, "installd not found; trying again");
-            BackgroundThread.getHandler().postDelayed(() -> {
-                connect();
-            }, CONNECT_RETRY_DELAY_MS);
+            new Thread(new Runnable() {
+                public void run() {
+                    try {
+                        Thread.sleep(CONNECT_RETRY_DELAY_MS);
+                        connect();
+                    } catch (Exception e) {
+                        Slog.e(TAG, "found exception " + e + " in reconnect()");
+                    }
+                }
+            }).start();
         }
     }

--
2.34.1

